### DISCUSSION TITLE
Orange Pi 5 Hardware acceleration

### DISCUSSION THREAD

**Discussion by stallion25** [Joshua-Riek/ubuntu-rockchip#115] [Q&A]
I saw newest builds for my orange pi.

Are those builds supporting hardware acceleration (Mali) in Orange pi 5?

What are glmark2(-es etc) scores for those builds?

Are onborads devices working like 1,3,5 I2C's, UART's  etc?

regards

Tom

**Comment by Joshua-Riek** [comment]
The desktop builds support hardware acceleration, but the server builds do not.

The scores for glmark2 can vary on what the system is doing in the background. However, these are the average numbers I found:
- glmark2: 2000
- gmlark2-es2: 2000
- glmark2-es2-wayland: 4700

Onboard devices work perfectly, such as SPI, I2C, and UART. 
I have a simple example below of how to enable individual devices using the included device tree overlays:

```bash
# Enable i2c
echo "i2c1-m2" | sudo tee -a /boot/firmware/overlays.txt

# Enable spidev
echo "spi4-m0-cs1-spidev" | sudo tee -a /boot/firmware/overlays.txt
```

**Reply by stallion25** [reply]
According to my last 2 weeks trials, its fastest build with 3d aceeleration I had so far. Thankyou !

I still cannot get 1,3,5 i2c's working:

root@op5:~# i2cdetect -l
i2c-0   i2c             rk3x-i2c                                I2C adapter
i2c-2   i2c             rk3x-i2c                                I2C adapter
i2c-6   i2c             rk3x-i2c                                I2C adapter
i2c-7   i2c             rk3x-i2c                                I2C adapter
i2c-9   i2c             fde50000.dp                             I2C adapter
i2c-10  i2c             ddc                                     I2C adapter
root@op5:~# cat /boot/firmware/overlays.txt
overlays=
i2c1-m4
i2c3-m0
i2c5-m3
root@op5:~#

can anyone help ?

**Reply by Joshua-Riek** [reply]
overlays.txt should look like this:
```
overlays=i2c1-m4 i2c3-m0 i2c5-m3
```
Then reboot your system.

**Reply by stallion25** [reply]
Its working now, thanks a lot for your work Joshua!

Last question, where is configured overlay , I mean where is orange pi 5 overlay chosen instead of orange pi 5B ?

**Reply by Joshua-Riek** [reply]
Right now, it's in the uboot script under the system path /boot/firmware/boot.cmd. It should be on line 8 of boot.cmd.
```bash
load ${devtype} ${devnum}:${distro_bootpart} ${fdt_addr_r} /rk3588s-orangepi-5.dtb
```
If you make any modifications to boot.cmd you will need to recompile the script.
```bash
sudo mkimage -A arm64 -O linux -T script -C none -n "Boot Script" -d boot.cmd boot.scr
```
Side note: I have been making changes to remove the /boot/firmware/ path and put everything under /boot/, but this will be in the future.

**Reply by stallion25** [reply]
That works awesome. 

I was trying several wifi dongles  with this board.

only realtek one is visible as wlan device

both rt0x200 cards are in lsusb and  correct kerlnel driver rt2x00usb is loaded.

below some more info:

root@op5:~# lsusb
Bus 006 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 008 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 007 Device 002: ID 413c:301d Dell Computer Corp. Dell Universal Receiver
Bus 007 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 004 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 002 Device 006: ID 0bda:8176 Realtek Semiconductor Corp. RTL8188CUS 802.11n WLAN Adapter
Bus 002 Device 004: ID 148f:3070 Ralink Technology, Corp. RT2870/RT3070 Wireless Adapter
Bus 002 Device 003: ID 148f:3572 Ralink Technology, Corp. RT3572 Wireless Adapter
Bus 002 Device 002: ID 214b:7250 Huasheng Electronics USB2.0 HUB
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 003 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
root@op5:~# ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.254.10  netmask 255.255.255.0  broadcast 192.168.254.255
        inet6 fe80::c2c3:ce38:c537:296d  prefixlen 64  scopeid 0x20<link>
        ether 1a:17:97:fb:e4:4b  txqueuelen 1000  (Ethernet)
        RX packets 938  bytes 161570 (161.5 KB)
        RX errors 0  dropped 70  overruns 0  frame 0
        TX packets 139  bytes 29229 (29.2 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 86

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 178  bytes 16828 (16.8 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 178  bytes 16828 (16.8 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlx00e04d390a39: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.8.133  netmask 255.255.255.0  broadcast 192.168.8.255
        inet6 fe80::40e7:ce77:5b17:cbc6  prefixlen 64  scopeid 0x20<link>
        ether 00:e0:4d:39:0a:39  txqueuelen 1000  (Ethernet)
        RX packets 1477  bytes 303809 (303.8 KB)
        RX errors 0  dropped 69  overruns 0  frame 0
        TX packets 666  bytes 463186 (463.1 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

what can be a problem ?

**Reply by stallion25** [reply]
Found something interesting, looks like rt2800usb and rt2800lib modules are not compiled in this kernel:

root@op5:/home/stallion# ls /lib/modules/5.10.110-rockchip-rk3588/kernel/drivers                                                                             /net/wireless/ralink/rt2x00/
rt2500usb.ko  rt2x00lib.ko  rt2x00usb.ko

any idea how to compile them ?

**Reply by Joshua-Riek** [reply]
I see the below options in the kernel config. However, it would require a re-compile, are you familiar with compiling the kernel at all?

```
joshua@GS66-Stealth-11UE:~/Documents/ubuntu-orange-pi5$ cat build/linux-orangepi/.config | grep -i rt2
CONFIG_RT2X00=m
# CONFIG_RT2400PCI is not set
# CONFIG_RT2500PCI is not set
# CONFIG_RT2800PCI is not set
CONFIG_RT2500USB=m
# CONFIG_RT2800USB is not set
CONFIG_RT2X00_LIB_USB=m
CONFIG_RT2X00_LIB=m
CONFIG_RT2X00_LIB_CRYPTO=y
CONFIG_RT2X00_LIB_LEDS=y
# CONFIG_RT2X00_LIB_DEBUGFS is not set
# CONFIG_RT2X00_DEBUG is not set
# CONFIG_SERIAL_8250_RT288X is not set
```

**Reply by stallion25** [reply]
i changed rt2x to modules in krenel .config , when trying recompile kernel im getting error:

fatal error: classmap.h: No such file or directory
   18 | #include "classmap.h"
      | ^~~~~~~~~~~~
compilation terminated.

found this problem on very old forums with 3.x kernel, but 5.x ??

was using fresh image TF for it and kernel building apt update

how do you compiled kernel for thoose ubuntu builds ?

regards

Tom

**Reply by Joshua-Riek** [reply]
I'm making changes to have this more straightforward. However, it's a little complicated at the moment.

```bash
# Install packages
sudo apt-get install -y build-essential gcc-aarch64-linux-gnu bison \
qemu-user-static qemu-system-arm qemu-efi u-boot-tools binfmt-support \
debootstrap flex libssl-dev bc rsync kmod cpio xz-utils fakeroot parted \
udev dosfstools uuid-runtime git-lfs device-tree-compiler python2 python3 \
python-is-python3 fdisk

# Clone git repo
git clone https://github.com/Joshua-Riek/ubuntu-orange-pi5.git
cd ubuntu-orange-pi5

# Download and patch the orange pi linux kernel source
if [ ! -d linux-orangepi ]; then
    git clone --progress -b orange-pi-5.10-rk3588 https://github.com/orangepi-xunlong/linux-orangepi.git
    git -C linux-orangepi checkout 0d1781e72fb7707a0bbc4419c8f1bc75a113f19a
    for patch in ../patches/linux-orangepi/*.patch; do git -C linux-orangepi apply "../${patch}"; done
fi
cd linux-orangepi

mkdir -p build && cd build

# Set kernel config 
cp ../../config/linux-rockchip-rk3588-legacy.config .config
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- olddefconfig
./scripts/config --disable CONFIG_DEBUG_INFO
./scripts/config --disable CONFIG_MODULE_SCMVERSION

touch .version

# Compile kernel into deb package
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- KBUILD_DEBARCH=arm64 KDEB_PKGVERSION="5.10.110-1" LOCALVERSION="-rockchip-rk3588" -j "$(nproc)" bindeb-pkg
```

**Reply by stallion25** [reply]
I was using img builder https://github.com/Joshua-Riek/ubuntu-orange-pi5

i get something like that:

tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/firmware/dtbs/overlays/rk3588-lcd2.dtbo: Cannot open: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/firmware/dtbs/overlays/rk3588-dmc.dtbo: Cannot open: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/firmware/dtbs/overlays/rk3588-uart0-m2.dtbo: Cannot open: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/firmware/dtbs/overlays/rk3588-pwm13-m2.dtbo: Cannot open: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/firmware/dtbs/overlays/rk3588-i2c5-m3.dtbo: Cannot open: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/initrd.img: Cannot create symlink to ‘initrd.img-5.10.110-orangepi-rk3588’: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/System.map: Cannot create symlink to ‘System.map-5.10.110-orangepi-rk3588’: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/vmlinuz: Cannot create symlink to ‘vmlinuz-5.10.110-orangepi-rk3588’: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/vmlinuz-5.10.110-orangepi-rk3588: Cannot open: No such file or directory
tar: ./boot: Cannot mkdir: No space left on device
tar: ./boot/config-5.10.110-orangepi-rk3588: Cannot open: No such file or directory
tar: Exiting with failure status due to previous errors
Error: in ../scripts/build-image.sh on line 123
Error: in ./scripts/build-rootfs.sh on line 276
Error: in ./build.sh on line 102
root@opi5:/usr/src/ubuntu-orange-pi5#

tmp device at those time look like out of space:

root@opi5:~# df
Filesystem     1K-blocks     Used Available Use% Mounted on
tmpfs            1609628     4472   1605156   1% /run
/dev/mmcblk1p2 240132824 22351844 207913656  10% /
tmpfs            8048136        0   8048136   0% /dev/shm
tmpfs               5120        4      5116   1% /run/lock
tmpfs               4096        0      4096   0% /sys/fs/cgroup
tmpfs            8048136        0   8048136   0% /run/qemu
/dev/mmcblk1p1    261868      600    261268   1% /boot/firmware
tmpfs            1609624      112   1609512   1% /run/user/1000
/dev/loop0p1      261864        0    261864   0% /tmp/mnt/system-boot
/dev/loop0p2     2236672  2056048     46944  98% /tmp/mnt/writable


were you using 32GB version to build ?

regards
Tomasz

**Reply by Joshua-Riek** [reply]
Oh, my bad. I forgot you need to use a different tag of the source code. Delete the ubuntu-orange-pi5 folder and clone the repo again like below.
Your boot partition might need to be cleaned up a little, could you show me the `ls -la` output of /boot/firmware?
```
git clone https://github.com/Joshua-Riek/ubuntu-orange-pi5.git
cd ubuntu-orange-pi5
git checkout v1.12
```

**Reply by stallion25** [reply]
root@opi5:~# cd /boot/firmware
root@opi5:/boot/firmware# ls -la
total 500
drwxr-xr-x 3 root root  16384 sty  1  1970 .
drwxr-xr-x 3 root root   4096 kwi 16 14:10 ..
-rwxr-xr-x 1 root root   1555 kwi 12 00:16 boot.cmd
-rwxr-xr-x 1 root root   1627 kwi 12 00:16 boot.scr
drwxr-xr-x 2 root root   4096 kwi 11 23:50 overlays
-rwxr-xr-x 1 root root     10 kwi 12 00:16 overlays.txt
-rwxr-xr-x 1 root root 233724 kwi 11 23:42 rk3588s-orangepi-5b.dtb
-rwxr-xr-x 1 root root 234473 kwi 11 23:42 rk3588s-orangepi-5.dtb

nothing much.
Im using 256 GB TF and 256 GB nvme

im not changing 16MB partition

I just cat TF to NVME and change UUID: 

root@opi5:~# mount
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
udev on /dev type devtmpfs (rw,nosuid,relatime,size=8000260k,nr_inodes=2000065,mode=755)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=1609628k,mode=755)
/dev/mmcblk1p2 on / type ext4 (rw,relatime)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev)
tmpfs on /run/lock type tmpfs (rw,nosuid,nodev,noexec,relatime,size=5120k)
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,size=4096k,nr_inodes=1024,mode=755)
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,name=systemd)
pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
none on /sys/fs/bpf type bpf (rw,nosuid,nodev,noexec,relatime,mode=700)
cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,nosuid,nodev,noexec,relatime,hugetlb)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,nosuid,nodev,noexec,relatime,perf_event)
cgroup on /sys/fs/cgroup/pids type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpu,cpuacct)
cgroup on /sys/fs/cgroup/net_cls,net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls,net_prio)
cgroup on /sys/fs/cgroup/rdma type cgroup (rw,nosuid,nodev,noexec,relatime,rdma)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=15729)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime,pagesize=2M)
debugfs on /sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime)
tracefs on /sys/kernel/tracing type tracefs (rw,nosuid,nodev,noexec,relatime)
fusectl on /sys/fs/fuse/connections type fusectl (rw,nosuid,nodev,noexec,relatime)
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
none on /run/credentials/systemd-sysusers.service type ramfs (ro,nosuid,nodev,noexec,relatime,mode=700)
tmpfs on /run/qemu type tmpfs (rw,nosuid,nodev,relatime,mode=755)
/dev/mmcblk1p1 on /boot/firmware type vfat (rw,relatime,fmask=0022,dmask=0022,codepage=936,iocharset=utf8,shortname=mixed,utf8,errors=remount-ro)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,nosuid,nodev,noexec,relatime)
tmpfs on /run/user/1000 type tmpfs (rw,nosuid,nodev,relatime,size=1609624k,nr_inodes=402406,mode=700,uid=1000,gid=1000)

**Comment by stallion25** [comment]
I had also found xrdp is always logging into default ,empty XOrg login screen instead of Gnome.

Can you check it?
